src/dma.md

```diff
@@ -0,0 +1,348 @@
+variables `x` and `y` changing their value at random times. The DMA transfer
+could also overwrite the state (e.g. link register) pushed onto the stack by the
+prologue of function `bar`.
+
+Note that if we had not use `mem::forget`, but `mem::drop`, it would have been
+possible to make `Transfer`'s destructor stop the DMA transfer and then the
+program would have been safe. But one can *not* rely on destructors running to
+enforce memory safety because `mem::forget` and memory leaks (see RC cycles) are
+safe in Rust.
+
+We can fix this particular problem by changing the lifetime of the buffer from
+`'a` to `'static` in both APIs.
+
+``` rust
+{{#include ../ci/dma/examples/two.rs:7:12}}
+{{#include ../ci/dma/examples/two.rs:21:27}}
+{{#include ../ci/dma/examples/two.rs:35:36}}
+```
+
+If we try to replicate the previous problem we note that `mem::forget` no longer
+causes problems.
+
+``` rust
+{{#include ../ci/dma/examples/two.rs:40:52}}
+{{#include ../ci/dma/examples/two.rs:54:61}}
+```
+
+As before, the DMA transfer continues after `mem::forget`-ing the `Transfer`
+value. This time that's not an issue because `buf` is statically allocated
+(e.g. `static mut` variable) and not on the stack.
+
+## Overlapping use
+
+Our API doesn't prevent the user from using the `Serial` interface while the DMA
+transfer is in progress. This could lead the transfer to fail or data to be
+lost.
+
+There are several ways to prevent overlapping use. One way is to have `Transfer`
+take ownership of `Serial1` and return it back when `wait` is called.
+
+``` rust
+{{#include ../ci/dma/examples/three.rs:7:32}}
+{{#include ../ci/dma/examples/three.rs:40:53}}
+{{#include ../ci/dma/examples/three.rs:60:68}}
+```
+The move semantics statically prevent access to `Serial1` while the transfer is
+in progress.
+
+``` rust
+{{#include ../ci/dma/examples/three.rs:71:81}}
+```
+
+There are other ways to prevent overlapping use. For example, a (`Cell`) flag
+that indicates whether a DMA transfer is in progress could be added to
+`Serial1`. When the flag is set `read`, `write`, `read_exact` and `write_all`
+would all return an error (e.g. `Error::InUse`) at runtime. The flag would be
+set when `write_all` / `read_exact` is used and cleared in `Transfer.wait`.
+
+## Compiler (mis)optimizations
+
+The compiler is free to re-order and merge non-volatile memory operations to
+better optimize a program. With our current API, this freedom can lead to
+undefined behavior. Consider the following example:
+
+``` rust
+{{#include ../ci/dma/examples/three.rs:84:97}}
+```
+
+Here the compiler is free to move `buf.reverse()` before `t.wait()`, which would
+result in a data race: both the processor and the DMA would end up modifying
+`buf` at the same time. Similarly the compiler can move the zeroing operation to
+after `read_exact`, which would also result in a data race.
+
+To prevent these problematic reorderings we can use a [`compiler_fence`]
```

Comment:
In the examples, should the fences not be reversed?
Not 100% sure myself, but when the buffer is given to the DMA is it not acquired?

Reply:
The next five paragraphs explain why these orderings are used. Can you elaborate on why you think the orderings should be reversed? (Also, don't take the names "acquire" and "release" literally; read the description of their semantics in the linked documentation)

Reply:
Yeah, no, I was doing this a lot with my SPSC queue in C++ some time ago and the semantics seems the same unless I am missing some detail:

* `acquire` when you are accessing data/buffer that needs to be synced 
* `release` after you have finished writing to a buffer or modifying a buffer

But it should be fine, I was thinking of it as buffer access, and you are "done" with, for example, the pointer which results in a `release`, so I was in error and this is fine.